### PR TITLE
refactor: add a new function to store app_info into file

### PR DESCRIPTION
**Pull request by hycdong**

### CODE REVIEW

File: src/replica/replica.cpp [hunk 1]
```diff
@@ -556,5 +557,16 @@ void replica::init_disk_tag()
     }
 }
 
+error_code replica::store_app_info(app_info info, const std::string &path)
```
Comment (levy5307) [inline src/replica/replica.cpp:560]: app_info &info

File: src/replica/replica.cpp [hunk 2]
```diff
@@ -556,5 +557,16 @@ void replica::init_disk_tag()
     }
 }
 
+error_code replica::store_app_info(app_info &info, const std::string &path)
+{
+    replica_app_info new_info((app_info *)&info);
+    auto info_path = path.empty() ? utils::filesystem::path_combine(_dir, ".app-info") : path;
```
Comment (levy5307) [inline src/replica/replica.cpp:563]: const auto &info_path

File: src/replica/replica.h [hunk 3]
```diff
@@ -451,6 +451,9 @@ class replica : public serverlet<replica>, public ref_counter, public replica_ba
 
     void init_disk_tag();
 
+    // store `ainfo` into `path` directory
```
Comment (levy5307) [inline src/replica/replica.h:454]: Could you add some comments here to show what would happen if path = ""?

File: src/replica/replica.cpp [hunk 4]
```diff
@@ -556,5 +557,17 @@ void replica::init_disk_tag()
     }
 }
 
+error_code replica::store_app_info(app_info &info, const std::string &path)
+{
+    replica_app_info new_info((app_info *)&info);
+    const auto &info_path =
+        path.empty() ? utils::filesystem::path_combine(_dir, ".app-info") : path;
+    auto err = new_info.store(info_path.c_str());
+    if (err != ERR_OK) {
```
Comment (levy5307) [inline src/replica/replica.cpp:566]: I recommend to add dsn_unlikely if err != ERR_OK is rarely happen

**Review by levy5307 — APPROVED** [review]

File: src/replica/replica.cpp [hunk 5]
```diff
@@ -556,5 +557,17 @@ void replica::init_disk_tag()
     }
 }
 
+error_code replica::store_app_info(app_info &info, const std::string &path)
+{
+    replica_app_info new_info((app_info *)&info);
+    const auto &info_path =
+        path.empty() ? utils::filesystem::path_combine(_dir, ".app-info") : path;
```
Comment (Shuo-Jia) [inline src/replica/replica.cpp:564]: suggest assign `str` to `const` variable: kAppInfo = ".app-info";

I had added variable: https://github.com/XiaoMi/rdsn/blob/982c2bd180a081fe2c772191cad0ea279e0e2685/src/replica/replica_disk_migrator.cpp#L35

you can refator it into `replica.h`

Comment (hycdong) [reply src/replica/replica.cpp]: Good idea, code updated.

**Review by Shuo-Jia — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Pull request merged**